das/algorithms/3DES.py

```diff
@@ -0,0 +1,324 @@
+    for i in range(len(s)):
+        res = res + map_of_h2b[s[i]]
+
+    return res
+
+
+def bin2hex(s):
+    h = {"0000": '0',
+                  "0001": '1',
+                  "0010": '2',
+                  "0011": '3',
+                  "0100": '4',
+                  "0101": '5',
+                  "0110": '6',
+                  "0111": '7',
+                  "1000": '8',
+                  "1001": '9',
+                  "1010": 'A',
+                  "1011": 'B',
+                  "1100": 'C',
+                  "1101": 'D',
+                  "1110": 'E',
+                  "1111": 'F'}
+
+    res = "".join(h[s[i:i+4]] for i in range(0, len(s), 4))
+    return res
+
+
+def LS(bit_string, num_of_shifts):
+    s = ""
+    for i in range(num_of_shifts):
+        for j in range(1, len(bit_string)):
+            s = s + bit_string[j]
+        s = s + bit_string[0]
+        bit_string = s
+        s = ""
+    return bit_string
+
+
+def RS(bit_string, num_of_shifts):
+    for i in range(num_of_shifts):
+        bit_string = bit_string[-1] + bit_string[:-1]
+
+    return bit_string
+
+
+def xor(a, b):
+    result = ""
+    for i in range(len(a)):
+        if a[i] == b[i]:
+            result = result + "0"
+        else:
+            result = result + "1"
+    return result
+
+
+def get_round_key_encrypt(key):
+    round_keys = []
+
+    # PC-1
+    key = "".join(key[i-1] for i in keyp)
+
+    # C0
+    left_key = key[:28]
+    # D0
+    right_key = key[28:]
+    for round_count in range(16):
+        left_key = LS(left_key, shift_table[round_count])
+        right_key = LS(right_key, shift_table[round_count])
+
+        k = left_key + right_key
+        round_keys.append("".join(k[i-1] for i in key_comp))
+
+    return round_keys
+
+
+def get_round_keys_decrypt(key):
+    round_keys = []
+
+    # PC-1
+    key = "".join(key[i-1] for i in keyp)
+
+    # C0
+    left_key = key[:28]
+    # D0
+    right_key = key[28:]
+    for round_count in range(15,-1,-1):
+        k = left_key + right_key
+        round_keys.append("".join(k[i-1] for i in key_comp))
+
+        left_key = RS(left_key, shift_table[round_count])
+        right_key = RS(right_key, shift_table[round_count])
+    return round_keys
+
+
+def Expansion(binary):
+    return "".join([binary[i-1] for i in exp_d])
+
+
+def Permutation(binary):
+    return "".join([binary[i-1] for i in per])
+
+
+def S(binary):
+    inputs = [binary[i:i+6] for i in range(0,48,6)]
+    coords = [(bin2dec(x[0]+x[-1]), bin2dec(x[1:-1])) for x in inputs]
+    outputs = [dec2bin(sbox[i][x[0]][x[1]]) for i, x in enumerate(coords)]
+    return "".join(outputs)
+
+
+def F(right, k):
+    tmp = Expansion(right)
+    tmp = xor(tmp, k)
+    tmp = S(tmp)
+    tmp = Permutation(tmp)
+    return tmp
+
+
+def encrypt(pt, key):
+    assert len(pt) == 64
+
+    left_pt = pt[0:32]
+    right_pt = pt[32:]
+
+    for k in get_round_key_encrypt(key):
+        left_pt, right_pt = right_pt, xor(left_pt, F(right_pt, k))
```

Comment:
AFAIK this is functionally identical to the code you had within this for-loop, so I don't think there were any issues there.

Reply:
same goes for the for-loop in `decrypt`